pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java

```diff
@@ -1011,8 +1011,12 @@ private static String createJobName(Function.FunctionDetails functionDetails) {
                 functionDetails.getName());
     }
 
+    private static String toValidPodName(String ori) {
+        return ori.toLowerCase().replaceAll("[^a-z0-9]", "-");
+    }
+
     private static String createJobName(String tenant, String namespace, String functionName) {
-        return "pf-" + tenant + "-" + namespace + "-" + functionName;
+        return "pf-" + toValidPodName(tenant) + "-" + toValidPodName(namespace) + "-" + toValidPodName(functionName);
```

Comment:
This is a potentially breaking change. For example the "." character is valid and a user might already have functions running with a "." in the names. If they update the worker with this change the function apis will no longer work for those functions. Also, this could potentially cause naming collisions. The current way of naming pods is not great but I think there can be a better solutions to this and we need to maintain backwards compatibility in any changes like these. One idea is to use the UUID of the function. This will avoid naming collisions and be a valid pod name. The downside to this is that it makes it more difficult to identify function pods from their name.

Reply:
@cckellogg The intension of this PR is to solve problem if `functionName` not meet the `VALID_POD_NAME_REGEX` will cause broker pod panic. Once all brokers are assigned to this function, the broker cluster will down. 
- For backwards compatibility, i do forgot to consider in this PR, but the backwards compatibility is based on those `functionName` already match the `VALID_POD_NAME_REGEX`. Thus, we can change the logic here to:
  - test if `"pf-" + tenant + "-" + namespace + "-" + functionName` returns valid pod name which matches `VALID_POD_NAME_REGEX`.
  - if match, we use it.
  - if not match, we use `toValidPodName` in this PR to generate a valid pod name with `toLowerCase()` and mapping non-valid chars to "-".

- For "." problem, i can change the regex part to not map "." to "-". 
```
    private static String toValidPodName(String ori) {
        return ori.toLowerCase().replaceAll("[^a-z0-9-\.]", "-");
    }
```

- For naming collisions, how about to append a short hash to the pod name?

Reply:
To maintain backward compatibility a hash should only be added in certain cases. Also, the name needs to be predictable because there is a service that gets created and it's how the current implementation communicates with the function pods. 

The logic could be something like this

```
private static String createJobName(String tenant, String namespace, String functionName) {
  final String jobName = "pf-" + tenant + "-" + namespace + "-" + functionName;
  final String convertedJobName = toValidPodName(jobName);
  // we have a valid name based on the tenant/namespace/function
  if (jobName.equals.(convertedJobName) {
     return jobName;
  }
  // the fqn of the function contains characters not accept in kubernetes
  // maybe add a hash here to avoid potential naming collisions 
  // whatever gets added to avoid the collision it needs to always be the same for that fqn
  return convertedJobName
}
```

Also, can we add some unit tests for this we don't want to break things for users if they update their cluster.